go/oppobloom/oppobloom.go

```diff
@@ -36,16 +38,49 @@ func NewFilter(size int) (*Filter, error) {
 	size = int(math.Pow(2, math.Ceil(math.Log2(float64(size)))))
 	slice := make([]*[]byte, size)
 	sizeMask := uint32(size - 1)
-	return &Filter{slice, sizeMask}, nil
+	return &Filter{array: slice, sizeMask: sizeMask}, nil
 }
 
+// Adds the given bytes to the set, and indicates if they were already present in the set.
+// A true value here is definitive; a false value may be a false negative.
 func (f *Filter) Contains(id []byte) bool {
+	ret, _ := f.ContainsCollision(id)
+	return ret
+}
+
+// Like Contains, but also indicates if there was a collision on the key. If both are false,
+// then you can be sure that it is not a false negative. It may also be interested to track
+// how often collisons happen-- that tracking is left to external concerns.
+func (f *Filter) ContainsCollision(id []byte) (contains bool, collision bool) {
 	h := md5UintHash{md5.New()}
 	h.Write(id)
 	uindex := h.Sum32() & f.sizeMask
 	index := int32(uindex)
 	oldId := getAndSet(f.array, index, id)
-	return bytes.Equal(oldId, id)
+	contains = bytes.Equal(oldId, id)
+	collision = len(oldId) != 0 && !contains
+	if !contains && !collision {
+		atomic.AddUint32(&f.numEntries, 1)
```

Comment:
Hm, all this code from here and below seems like it could live on a struct that wraps around this one if the ContainsCollision method was added.

I'd be more inclined to add just that method and skip the rest of this code. In part, for better separation of concerns, but also because I don't think the numEntries code is quite what a lot of people need. (It seems to count the number of times a new entry was put in, sure, but its name suggests the number of items in the map at any given time)

Reply:
`numEntries` is actually meant to count the number of entries in the map-- it won't be incremented if there is a collision (in that case, there wouldn't be a new item added to the map, it would be replacing an item in the map).

Reply:
Good idea on the wrapper struct. I just pushed this up, if you'd like to take another look.